File: src/2025-02-22-00-51-50--a-deepseek-r1-test-220-10-1--2-1-huoshan/javascript/exercises/practice/two-bucket/two-bucket.js
//
// This is only a SKELETON file for the 'Two Bucket' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export class TwoBucket {
  constructor(bucket1Size, bucket2Size, goal, startBucket) {
    if (goal > Math.max(bucket1Size, bucket2Size)) {
      throw new Error('Goal cannot be larger than both buckets');
    }
    
    const gcdVal = TwoBucket.gcd(bucket1Size, bucket2Size);
    if (goal !== bucket1Size && goal !== bucket2Size && goal % gcdVal !== 0) {
      throw new Error('No solution possible');
    }

    this.bucket1Size = bucket1Size;
    this.bucket2Size = bucket2Size;
    this.goal = goal;
    this.startBucket = startBucket;
  }

  static gcd(a, b) {
    while (b !== 0) {
      [a, b] = [b, a % b];
    }
    return a;
  }

  solve() {
    const queue = [];
    const visited = new Set();

    // Initialize based on starting bucket
    let initialState;
    if (this.startBucket === 'one') {
      initialState = { b1: this.bucket1Size, b2: 0, moves: 1 };
    } else {
      initialState = { b1: 0, b2: this.bucket2Size, moves: 1 };
    }

    // Check initial state immediately
    if (initialState.b1 === this.goal) {
      return {
        moves: initialState.moves,
        goalBucket: 'one',
        otherBucket: initialState.b2
      };
    }
    if (initialState.b2 === this.goal) {
      return {
        moves: initialState.moves,
        goalBucket: 'two',
        otherBucket: initialState.b1
      };
    }

    queue.push(initialState);
    visited.add(`${initialState.b1},${initialState.b2}`);

    while (queue.length > 0) {
      const current = queue.shift();

      for (const next of this.generateNextStates(current)) {
        const stateKey = `${next.b1},${next.b2}`;
        
        if (visited.has(stateKey)) continue;
        if (this.isForbiddenState(next)) continue;

        // Check if we've reached the goal
        if (next.b1 === this.goal) {
          return {
            moves: next.moves,
            goalBucket: 'one',
            otherBucket: next.b2
          };
        }
        if (next.b2 === this.goal) {
          return {
            moves: next.moves,
            goalBucket: 'two',
            otherBucket: next.b1
          };
        }

        visited.add(stateKey);
        queue.push(next);
      }
    }

    throw new Error('No solution found');
  }

  generateNextStates(current) {
    const { b1, b2, moves } = current;
    const nextStates = [];

    // Fill operations
    if (b1 < this.bucket1Size) {
      nextStates.push({ b1: this.bucket1Size, b2, moves: moves + 1 });
    }
    if (b2 < this.bucket2Size) {
      nextStates.push({ b1, b2: this.bucket2Size, moves: moves + 1 });
    }

    // Empty operations
    if (b1 > 0) {
      nextStates.push({ b1: 0, b2, moves: moves + 1 });
    }
    if (b2 > 0) {
      nextStates.push({ b1, b2: 0, moves: moves + 1 });
    }

    // Pour operations
    const pour1to2 = Math.min(b1, this.bucket2Size - b2);
    if (pour1to2 > 0) {
      nextStates.push({
        b1: b1 - pour1to2,
        b2: b2 + pour1to2,
        moves: moves + 1
      });
    }

    const pour2to1 = Math.min(b2, this.bucket1Size - b1);
    if (pour2to1 > 0) {
      nextStates.push({
        b1: b1 + pour2to1,
        b2: b2 - pour2to1,
        moves: moves + 1
      });
    }

    return nextStates;
  }

  isForbiddenState(state) {
    if (this.startBucket === 'one') {
      return state.b1 === 0 && state.b2 === this.bucket2Size;
    }
    return state.b2 === 0 && state.b1 === this.bucket1Size;
  }
}

File: src/2025-02-22-00-51-50--a-deepseek-r1-test-220-10-1--2-1-huoshan/java/exercises/practice/two-bucket/src/main/java/TwoBucket.java
class TwoBucket {
    private int bucketOneCap;
    private int bucketTwoCap;
    private int desiredLiters;
    private String startBucket;
    private int moves;
    private String finalBucket;
    private int otherLiters;

    public TwoBucket(int bucketOneCap, int bucketTwoCap, int desiredLiters, String startBucket) {
        this.bucketOneCap = bucketOneCap;
        this.bucketTwoCap = bucketTwoCap;
        this.desiredLiters = desiredLiters;
        this.startBucket = startBucket;
        this.moves = 0;
        this.finalBucket = "";
        this.otherLiters = 0;
        
        solve();
    }

    private void solve() {
        // Check if initial fill meets the desired liters
        int initialB1 = startBucket.equals("one") ? bucketOneCap : 0;
        int initialB2 = startBucket.equals("two") ? bucketTwoCap : 0;
        moves = 1;
        
        if (initialB1 == desiredLiters) {
            finalBucket = "one";
            otherLiters = initialB2;
            return;
        } else if (initialB2 == desiredLiters) {
            finalBucket = "two";
            otherLiters = initialB1;
            return;
        }
        
        java.util.Queue<int[]> queue = new java.util.ArrayDeque<>();
        java.util.Set<String> visited = new java.util.HashSet<>();
        queue.add(new int[]{initialB1, initialB2, 1});
        visited.add(initialB1 + "," + initialB2);
        
        while (!queue.isEmpty()) {
            int[] current = queue.poll();
            int b1 = current[0];
            int b2 = current[1];
            int currentMoves = current[2];
            
            java.util.List<int[]> nextStates = new java.util.ArrayList<>();
            
            // Generate all possible next states
            nextStates.add(new int[]{0, b2, currentMoves + 1}); // Empty b1
            nextStates.add(new int[]{b1, 0, currentMoves + 1}); // Empty b2
            nextStates.add(new int[]{bucketOneCap, b2, currentMoves + 1}); // Fill b1
            nextStates.add(new int[]{b1, bucketTwoCap, currentMoves + 1}); // Fill b2
            
            // Pour b1 to b2
            int pourAmount = Math.min(b1, bucketTwoCap - b2);
            nextStates.add(new int[]{b1 - pourAmount, b2 + pourAmount, currentMoves + 1});
            
            // Pour b2 to b1
            pourAmount = Math.min(b2, bucketOneCap - b1);
            nextStates.add(new int[]{b1 + pourAmount, b2 - pourAmount, currentMoves + 1});
            
            for (int[] state : nextStates) {
                int newB1 = state[0];
                int newB2 = state[1];
                int newMoves = state[2];
                String stateKey = newB1 + "," + newB2;
                
                // Check if state is invalid (starting bucket empty and other full)
                boolean invalid = false;
                if (startBucket.equals("one")) {
                    if (newB1 == 0 && newB2 == bucketTwoCap) invalid = true;
                } else {
                    if (newB2 == 0 && newB1 == bucketOneCap) invalid = true;
                }
                
                if (invalid || visited.contains(stateKey)) continue;
                
                // Check if desired liters found
                if (newB1 == desiredLiters) {
                    moves = newMoves;
                    finalBucket = "one";
                    otherLiters = newB2;
                    return;
                } else if (newB2 == desiredLiters) {
                    moves = newMoves;
                    finalBucket = "two";
                    otherLiters = newB1;
                    return;
                }
                
                visited.add(stateKey);
                queue.add(state);
            }
        }
        
        throw new IllegalArgumentException("No solution possible");
    }

    int getTotalMoves() {
        return moves;
    }

    String getFinalBucket() {
        return finalBucket;
    }

    int getOtherBucket() {
        return otherLiters;
    }
}
